### PR TITLE
Georgr/refactor gap8

### PR DESCRIPTION
**Pull request by da-gazzi**
Refactor GAP8 backend, unifying the following code in a `Common` folder:
- Tilers
- Non-layer templates (makefile, network.c/h, etc)
- DMA and utility code
- Parsers
- HW Pattern rewriter
DMA driver is simplified and streamlined based on @lukamac 's code. 
The following keys were introduced in `HW_description.json`:
- `double_buffering`: Set to 2 for double-buffering, 1 for no double buffering (i.e., `GAP8_board` backend)
- `split_ints`: if `true`, checksums will be calculated by DORY as the sum of the uint8 representation of the results - even for int32 outputs (i.e., an int32 will be "split" into 4 uint8's)
- `blocking_dma_transfers`: if true, insert barriers after pushing 2D transfers in 3D and HWC-to-CHW modes (otherwise the GAP8 board freezes)
Tests are passing, some examples tested on GAPuino as well.

### CODE REVIEW

File: dory/Hardware_targets/GAP8/Common/Utils/dory_dma.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#elif defined(MCHAN_EVENT)
+#define MCHAN_FLAGS    (MCHAN_CMD_FLAG_EVENT_ENABLE | MCHAN_CMD_FLAG_INCREMENTAL)
+#elif defined(MCHAN_INTERRUPT)
+#define MCHAN_FLAGS    (MCHAN_CMD_FLAG_INTERRUPT_ENABLE | MCHAN_CMD_FLAG_INCREMENTAL)
+#endif
+
+#define MCHAN_FLAGS_1D (MCHAN_FLAGS)
+#define MCHAN_FLAGS_2D (MCHAN_FLAGS | MCHAN_CMD_FLAG_2D_TRANSFER_EXTERNAL)
+
+#define MIN(a,b) ((a)<(b)?(a):(b))
```
Comment (lukamac) [inline dory/Hardware_targets/GAP8/Common/Utils/dory_dma.h:50]: Could you move this to the `dory_dma.c` file and keep the header file as minimal as required?

File: dory/Hardware_targets/GAP8/Common/Utils/mchan.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+#else
+#define MCHAN_CMD_FLAG_DIRECTION_LOC2EXT    (0 << 16)
+#define MCHAN_CMD_FLAG_DIRECTION_EXT2LOC    (1 << 16)
+#define MCHAN_CMD_FLAG_INCREMENTAL          (1 << 17)
+#define MCHAN_CMD_FLAG_2D_TRANSFER_EXTERNAL (1 << 18)
+#define MCHAN_CMD_FLAG_EVENT_ENABLE         (1 << 19)
+#define MCHAN_CMD_FLAG_INTERRUPT_ENABLE     (1 << 20)
+#define MCHAN_CMD_FLAG_BROADCAST_FINISH     (1 << 21)
+#define MCHAN_CMD_SHIFT_DIRECTION (16)
+#endif
```
Comment (lukamac) [inline dory/Hardware_targets/GAP8/Common/Utils/mchan.h:54]: I would write this maybe like this:

```suggestion
// MCHAN version 7 has 1 more bit for the transfer length, so all the flag offsets are shifted by 1. Also, LOC (TCDM) striding is not supported in v6.
#if MCHAN_VERSION==7
#define MCHAN_CMD_TRANSFER_LENGTH_SIZE (17)
#else
#define MCHAN_CMD_TRANSFER_LENGTH_SIZE (16)
#endif

#define MCHAN_CMD_FLAG_DIRECTION_LOC2EXT    (0 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 0))
#define MCHAN_CMD_FLAG_DIRECTION_EXT2LOC    (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 0))
#define MCHAN_CMD_FLAG_INCREMENTAL          (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 1))
#define MCHAN_CMD_FLAG_2D_TRANSFER_EXTERNAL (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 2))
#define MCHAN_CMD_FLAG_EVENT_ENABLE         (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 3))
#define MCHAN_CMD_FLAG_INTERRUPT_ENABLE     (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 4))
#define MCHAN_CMD_FLAG_BROADCAST_FINISH     (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 5))
#if MCHAN_VERSION == 7
#define MCHAN_CMD_FLAG_2D_TRANSFER_LOCAL    (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 6))
#endif
#define MCHAN_CMD_SHIFT_DIRECTION (MCHAN_CMD_TRANSFER_LENGTH_SIZE)
```

Comment (lukamac) [reply dory/Hardware_targets/GAP8/Common/Utils/mchan.h]: @da-gazzi I think you forgot to add the flag for the local 2D transfer
```
#if MCHAN_VERSION == 7
#define MCHAN_CMD_FLAG_2D_TRANSFER_LOCAL    (1 << (MCHAN_CMD_TRANSFER_LENGTH_SIZE + 6))
#endif
```

File: dory/Hardware_targets/GAP8/Common/Utils/dory_dma.h (hunk 1 above)
Comment (da-gazzi) [reply dory/Hardware_targets/GAP8/Common/Utils/dory_dma.h]: done!

File: dory/Hardware_targets/GAP8/Common/Utils/mchan.h (hunk 2 above)
Comment (da-gazzi) [reply dory/Hardware_targets/GAP8/Common/Utils/mchan.h]: both fixed!

**Pull request merged**